lib/mix/lib/mix/project_stack.ex

```diff
@@ -263,4 +270,15 @@ defmodule Mix.ProjectStack do
     {reply, state} = fun.(state)
     {:reply, reply, state}
   end
+
+  @impl true
+  def handle_info({:compile_env, {app, path, return}}, {stack, post_config}) do
+    with [%{config: config, compile_env: compile_env} = project | stack] <- stack,
+         ^app <- config[:app] do
```

Comment:
Is this intended that if you read compile env from another application it's not recorded?

Reply:
Yes, I am not sure how it could work across applications, since this is stored in the `compile_env` of each individual application. I will add a note to `compile_env` docs saying we can only track for the current application and it is extremely discouraged to read another application's environment.

Reply:
If that's the case, should it produce some sort of warning when you read compile env from another application?
Alternatively, we could only allow reading env from the application currently being compiled, so it would always be `Application.compile_env([:foo, :bar, :baz])`

Reply:
The problem is that we don't have the application being compiled in Elixir. Everything is tracked by Mix since we have historically tried to keep these concerns separated (hence the listener approach

However, we may need to do further changes here. Today I have no idea which file called `compile_env`, so if that file no longer calls `compile_env`, I am unable to stop tracking it. I will have to make it a macro (but explicitly requiring Application is going to be a detriment to its usage). We could automatically require Application, but we avoided doing that as well.